url3426a97ee4_rtreffer_minidns_tgz-pJ8-de_measite_minidns_minidnsrepl_MiniDnsReplJ8/src/de/measite/minidns/dnssec/Verifier.java

```diff
@@ -0,0 +1,232 @@
+                    return new NSECDoesNotMatchReason(q, nsec3record);
+                }
+            }
+            return null;
+        }
+        if (nsecMatches(s, nsec3record.name.getHostpart(), Base32.encodeToString(nsec3.nextHashed))) {
+            return null;
+        }
+        return new NSECDoesNotMatchReason(q, nsec3record);
+    }
+
+    static byte[] combine(RRSIG rrsig, List<Record<? extends Data>> records) {
+        ByteArrayOutputStream bos = new ByteArrayOutputStream();
+        DataOutputStream dos = new DataOutputStream(bos);
+
+        // Write RRSIG without signature
+        try {
+            rrsig.writePartialSignature(dos);
+
+            DNSName sigName = records.get(0).name;
+            if (!sigName.isRootLabel()) {
+                if (sigName.getLabelCount() < rrsig.labels) {
+                    throw new DNSSECValidationFailedException("Invalid RRsig record");
+                }
+
+                if (sigName.getLabelCount() > rrsig.labels) {
+                    // Expand wildcards
+                    sigName = DNSName.from("*." + sigName.stripToLabels(rrsig.labels));
+                }
+            }
+
+            List<byte[]> recordBytes = new ArrayList<>();
+            for (Record<? extends Data> record : records) {
+                Record<Data> ref = new Record<>(sigName, record.type, record.clazzValue, rrsig.originalTtl, (Data) record.payloadData);
+                recordBytes.add(ref.toByteArray());
+            }
+
+            // Sort correctly (cause they might be ordered randomly)
+            final int offset = sigName.size() + 10; // Where the RDATA begins
+            Collections.sort(recordBytes, new Comparator<byte[]>() {
+                @Override
+                public int compare(byte[] b1, byte[] b2) {
+                    for (int i = offset; i < b1.length && i < b2.length; i++) {
+                        if (b1[i] != b2[i]) {
+                            return (b1[i] & 0xFF) - (b2[i] & 0xFF);
+                        }
+                    }
+                    return b1.length - b2.length;
+                }
+            });
+
+            for (byte[] recordByte : recordBytes) {
+                dos.write(recordByte);
+            }
+            dos.flush();
+        } catch (IOException e) {
+            // Never happens
+            throw new RuntimeException(e);
+        }
+        return bos.toByteArray();
+    }
+
+    static boolean nsecMatches(String test, String lowerBound, String upperBound) {
+        return nsecMatches(DNSName.from(test), DNSName.from(lowerBound), DNSName.from(upperBound));
+    }
+
+    /**
+     * Tests if a nsec domain name is part of an NSEC record.
+     *
+     * @param test       test domain name
+     * @param lowerBound inclusive lower bound
+     * @param upperBound exclusive upper bound
+     * @return test domain name is covered by NSEC record
+     */
+    static boolean nsecMatches(DNSName test, DNSName lowerBound, DNSName upperBound) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 11. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.